e2e-tests/src/main.rs

```diff
@@ -125,15 +118,81 @@ fn test_token_transfer(config: Config) -> anyhow::Result<()> {
     Ok(())
 }
 
+// todo: check channeling tips
+fn test_channeling_fee(config: Config) -> anyhow::Result<()> {
+    let (connection, _, to) = setup_for_transfer(config);
+    let treasury = get_treasury_account(&connection);
+
+    let treasury_balance_before = get_free_balance(&treasury, &connection);
+    let issuance_before = get_total_issuance(&connection);
+    info!(
+        "[+] Treasury balance before tx: {}. Total issuance: {}.",
+        treasury_balance_before, issuance_before
+    );
+
+    let tx = transfer(&to, 1000u128, &connection);
+
+    let treasury_balance_after = get_free_balance(&treasury, &connection);
+    let issuance_after = get_total_issuance(&connection);
+    info!(
+        "[+] Treasury balance after tx: {}. Total issuance: {}.",
+        treasury_balance_after, issuance_after
+    );
+
+    assert!(
+        issuance_after <= issuance_before,
+        "Unexpectedly {} was minted",
+        issuance_after - issuance_before
+    );
+    assert!(
+        issuance_before <= issuance_after,
+        "Unexpectedly {} was burned",
+        issuance_before - issuance_after
+    );
+
+    let fee_info = get_tx_fee_info(&connection, &tx);
+    let fee = fee_info.fee_without_weight + fee_info.adjusted_weight;
+
+    assert_eq!(
+        treasury_balance_before + fee,
+        treasury_balance_after,
+        "Incorrect amount was channeled to the treasury: before = {}, after = {}, fee = {}",
+        treasury_balance_before,
+        treasury_balance_after,
+        fee
+    );
+
+    Ok(())
+}
+
+fn test_treasury_access(config: Config) -> anyhow::Result<()> {
+    let Config { node, seeds, .. } = config.clone();
+
+    let proposer = get_first_account(&accounts(seeds));
+    let beneficiary = AccountId::from(proposer.public());
+    let connection = create_connection(node).set_signer(proposer);
+
+    propose_treasury_spend(0u128, &beneficiary, &connection);
+    propose_treasury_spend(0u128, &beneficiary, &connection);
+    let proposals_counter = get_proposals_counter(&connection);
+    assert!(proposals_counter >= 2, "Proposal was not created");
+
+    let sudo = get_sudo(config);
+    let connection = connection.set_signer(sudo);
+
+    send_treasury_approval(proposals_counter - 2, &connection);
+    wait_for_approval(&connection, proposals_counter - 2)?;
+
+    treasury_reject(proposals_counter - 1, &connection)?;
```

Comment:
This is a single function, but the above is two, confusing. Please make this consistent.

Reply:
I tried :P but unfortunately, these operations must be done in different ways :(
1. **approval** - sending an extrinsic is straightforward, but pallet_treasury does _not_ emit any corresponding event (apart from system.extrinsicSucceeded, but such event is in every block due to the timestamp extrinsic and there is no 'good' way of matching such result with the origin call); therefore, the only reliable way of verifying whether the approval was successful is just to look at the storage and check the `approvals` vec (and have a hope that the spending period has not just occurred :P)
2. **rejection** - again, sending an extrinsic is straightforward, and in this case there is a corresponding event, namely `Treasury::Rejected` so we can listen for this event; _however_, it often happens, that this event is dispatched before we start listening, so we need to setup listening before sending the extrinsic

Reply:
Ah, these are definitely annoyances.

To make this less confusing I would suggest creating a `treasury_approve` function, probably just consisting of the two lines above, at least then they look the same in the test and the reader doesn't have to spend time wondering why you only check the approval. Internally the functions will still differ, but the code will be more readable.

Reply:
done